test/jdk/tools/jpackage/share/JLinkOptionsTest.java

```diff
@@ -138,7 +139,44 @@ public void test() {
         }
     }
 
-    private final String[] required;
-    private final String[] prohibited;
-    private final JPackageCommand cmd;
+    @Test
+    public void testNoBindServicesByDefault() {
+        final var defaultModules = getModulesInRuntime();
+        final var modulesWithBindServices = getModulesInRuntime("--bind-services");
+
+        final var moduleComm = Comm.compare(defaultModules, modulesWithBindServices);
+
+        TKit.assertStringListEquals(List.of(), moduleComm.unique1().stream().toList(),
+                "Check '--bind-services' option doesn't remove modules");
+        TKit.assertNotEquals("", moduleComm.unique2().stream().sorted().collect(Collectors.joining(",")),
+                "Check '--bind-services' option adds modules");
+    }
+
+    private final JPackageCommand createJPackageCommand(String javaAppDesc) {
+        return JPackageCommand.helloAppImage(javaAppDesc).ignoreDefaultRuntime(true);
+    }
+
+    private final Set<String> getModulesInRuntime(String ... jlinkOptions) {
+        final var cmd = createJPackageCommand(PRINT_ENV_APP + "*");
+        if (jlinkOptions.length != 0) {
+            cmd.addArguments("--jlink-options");
+            cmd.addArguments(jlinkOptions);
+        }
+
+        cmd.executeAndAssertImageCreated();
+
+        final int attempts = 3;
+        final int waitBetweenAttemptsSeconds = 5;
+
+        final String output = configureEnvironment(new Executor())
+                .saveFirstLineOfOutput()
+                .setExecutable(cmd.appLauncherPath().toAbsolutePath())
+                .addArguments("--print-modules")
+                .executeAndRepeatUntilExitCode(0, attempts,
```

Comment:
Is there any reason why we need to repeat execution?

Reply:
Yes, launchers in tests are executed multiple times because they may sporadically crash due to an unidentified issue with JVM termination.

Usually, launchers are not explicitly executed in the tests. They are executed as a part of `JPackageCommand.executeAndAssertHelloAppImageCreated()` call, which attempts to run them multiple times if necessary.

In this test, launcher is explicitly executed and the setup to run it multiple times if needed is explicit.

Reply:
I reworked this part using `HelloApp.assertApp()` function. It wraps restart details.